### DISCUSSION TITLE
Does an Antlr4 g4 grammar exist for recent versions of C#

### DISCUSSION THREAD

**Discussion by DavidArno** [dotnet/csharplang#7379] [Q&A] [answered]
[There is a g4 grammar for C# on the antlr4 github repo](https://github.com/antlr/grammars-v4/blob/master/csharp/CSharpParser.g4), but this is out of date and doesn't eg have grammars for patterns. 

Does anyone know of a more up to date g4 grammar for C#?

**Comment by 333fred** [comment, marked as answer]
https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Generated/CSharp.Generated.g4

This is the only one I'm aware of, but it does have some problems when being considered for actual use:

* It's based on our syntax node structure, which is more permissive than C#'s actual grammar for error recovery purposes.
* We don't actually test it, so it's possible it has bugs or things that don't quite work in real antlr. We mainly just use it for code review of new syntax nodes.

**Reply by DavidArno** [reply]
That's perfect. I wanted it to play around with some "*what would C# look like if we designed it today*" ideas (with lots of my highly opinionated views influencing that of course 😃) so that's exactly the sort of thing I needed. Thanks.

**Reply by CyrusNajmabadi** [reply]
It's definitely legal antlr.  I validated it with antlr itself when I built the tool that creates it. Whether or not you can get a useful parser from our is another matter altogether.

**Reply by DavidArno** [reply]
What's the reasoning behind code like this being valid?
```cs
class Class1
{
    static void M()
    {
        cl\u0061ss.st\u0061tic(true);
    }
}
```
What benefit is there to supporting unicode escape sequences in identifiers? If you were "rebooting" C# and clearing out "undesirable" features (what's undesirable here is entirely subjective and opinionated), would unicode escape sequences in identifiers make the cut of features to keep? Does anyone use them?

**Reply by CyrusNajmabadi** [reply]
> What benefit is there to supporting unicode escape sequences in identifiers?

To ensure you can reference all names regardless of encoding/character/editor issues you might have.  It's helpful for when something is generating code and varying encoding systems may be complex and not easily controllable by the tool writing text.  This allows for always emitting as ascii while still having the full range of unicode be available.

> would unicode escape sequences in identifiers make the cut of features to keep? 

Probably yes.  They'd def be supported for things like strings.  And, at that point, not supporting them in all the other places where unicode is allowed would just be a tax.

> Does anyone use them?

Yes :)

**Reply by DreadLordMikey** [reply]
Apologize for the necro, but this entire thread concerns me directly for a project I'm working on right now.

Regarding Unicode characters in identifiers: We have many contractors we work with that are located in eastern European countries. Occasionally, I'll see them whip up quick code samples to share between themselves, and they are using their local region in Windows. The identifiers are ALWAYS using their local languages (Lithuanian and Ukranian) and character set. That requires Unicode support.

**Comment by The-Futurist** [comment]
> That's perfect. I wanted it to play around with some "_what would C# look like if we designed it today_" ideas (with lots of my highly opinionated views influencing that of course 😃) so that's exactly the sort of thing I needed. Thanks.

There's this too:

https://github.com/antlr/grammars-v4/tree/master/csharp

**Reply by CyrusNajmabadi** [reply]
Note: that grammar is roughly 8 years behind the language. It's fine for simple experimentation, but def would need major work to support modern c#.

**Reply by DavidArno** [reply]
@Korporal,

> There's this too:
> 
> https://github.com/antlr/grammars-v4/tree/master/csharp
> 

Yeah, I found that before, but it appeared out of date as it didn't have any pattern support for example. Thanks.
